qiita_client/plugin.py

```diff
@@ -0,0 +1,330 @@
+# -----------------------------------------------------------------------------
+
+import traceback
+import sys
+from string import ascii_letters, digits
+from random import SystemRandom
+from os.path import exists, join, expanduser
+from os import makedirs, environ
+from future import standard_library
+from json import dumps
+
+from qiita_client import QiitaClient
+
+with standard_library.hooks():
+    from configparser import ConfigParser
+
+
+class QiitaCommand(object):
+    """A plugin command
+
+    Parameters
+    ----------
+    name : str
+        The command name
+    description : str
+        The command description
+    function : callable
+        The function that executes the command. Should be a callable that
+        conforms to the signature:
+            `function(qclient, job_id, job_parameters, output_dir)`
+        where qclient is an instance of QiitaClient, job_id is a string with
+        the job identifier, job_parameters is a dictionary with the parameters
+        of the command and output_dir is a string with the output directory
+    required_parameters : dict of {str: (str, list of str)}
+        The required parameters of the command, keyed by parameter name. The
+        values should be a 2-tuple in which the first element is the parameter
+        type, and the second parameter is the list of subtypes (if applicable)
+    optional_parameters : dict of {str: (str, str)}
+        The optional parameters of the command, keyed by parameter name. The
+        values should be a 2-tuple in which the first element is the parameter
+        name, and the second paramater is the default value
+    outputs : dict of {str: str}
+        The description of the outputs that this command generated. The
+        format is: {output_name: artifact_type}
+    default_parameter_sets : dict of {str: dict of {str: str}}
+        The default parameter sets of the command, keyed by parameter set name.
+        The values should be a dictionary in which keys are the parameter names
+        and values are the specific value for each parameter
+
+    Raises
+    ------
+    TypeError
+        If `function` is not callable
+    ValueError
+        If `function` does not accept 4 parameters
+    """
+    def __init__(self, name, description, function, required_parameters,
+                 optional_parameters, outputs, default_parameter_sets=None):
+        self.name = name
+        self.description = description
+
+        # Make sure that `function` is callable
+        if not callable(function):
+            raise TypeError(
+                "Couldn't create command '%s': the provided function is not "
+                "callable (type: %s)" % (name, type(function)))
+
+        # `function` will be called with the following Parameters
+        # qclient, job_id, job_parameters, output_dir
+        # Make sure that `function` can receive 4 parameters
+        if function.__code__.co_argcount != 4:
```

Comment:
py2/py3 compatible. I want this utility library to be as general as possible so we can use it with QIIME1 and QIIME2, although I expect more devs use this to create the plugins of their own tools (which we don't control if they're py2 or py3).